cirq/contrib/acquaintance/gates.py

```diff
@@ -0,0 +1,311 @@
+            of the effect of inside-acquainting swaps.
+    """
+
+    max_reach = _get_max_reach(len(qubits), round_up=before)
+    reaches = chain(range(1, max_reach + 1),
+                    range(max_reach, -1, -1))
+    offsets = (0, 1) * max_reach
+    swap_gate = SwapPermutationGate(swap_gate)
+    ops = []
+    for offset, reach in zip(offsets, reaches):
+        if offset == before:
+            ops.append(acquaintance_gate)
+        for dr in range(offset, reach, 2):
+            ops.append(swap_gate(*qubits[dr:dr + 2]))
+    layers['pre' if before else 'post'] += ops
+
+    # add interstitial gate
+    interstitial = ('prior' if before else 'posterior') + '_interstitial'
+    layers[interstitial].append(acquaintance_gate)
+
+    # update mapping
+    reached_qubits = qubits[:max_reach + 1]
+    positions = list(mapping[q] for q in reached_qubits)
+    mapping.update(zip(reached_qubits, reversed(positions)))
+
+def _get_max_reach(size: int, round_up: bool=True) -> int:
+    if round_up:
+        return int(ceil(size / 2)) - 1
+    return max((size // 2) - 1, 0)
+
+
+def acquaint_and_shift(parts: Tuple[List[QubitId], List[QubitId]],
+                       layers: Dict[str, List[Operation]],
+                       acquaintance_size: int,
+                       swap_gate: Gate,
+                       mapping: Dict[QubitId, int]):
+    """Acquaints and shifts a pair of lists of qubits. The first part is
+    acquainted with every qubit individually in the second part, and vice
+    versa. Operations are grouped into several layers:
+        * prior_interstitial: The first layer of acquaintance gates.
+        * prior: The combination of acquaintance gates and swaps that acquaints
+            the inner halves.
+        * intra: The shift gate.
+        * post: The combination of acquaintance gates and swaps that acquaints
+            the outer halves.
+        * posterior_interstitial: The last layer of acquaintance gates.
+
+    Args:
+        parts: The two lists of qubits to acquaint.
+        layers: The layers to put gates into.
+        acquaintance_size: The number of qubits to acquaint at a time.
+        swap_gate: The gate used to swap logical indices.
+        mapping: The mapping from qubits to logical indices. Used to keep track
+            of the effect of inside-acquainting swaps.
+    """
+    left_part, right_part = parts
+    left_size, right_size = len(left_part), len(right_part)
+    assert not (set(left_part) & set(right_part))
+    qubits = left_part + right_part
+    shift = CircularShiftGate(left_size,
+                              swap_gate=swap_gate)(
+                                      *qubits)
+    if max(left_size, right_size) != acquaintance_size - 1:
+        layers['intra'].append(shift)
+    elif acquaintance_size == 2:
+        layers['prior_interstitial'].append(ACQUAINT(*qubits))
+        layers['intra'].append(shift)
+    else:
+        # before
```

Comment:
It should be possible to combine these four cases somehow, e.g. into a loop over two booleans. They have a ton of redundancy.